feat(guide): add guide for namespace scoped policies (#1910)
kumahq/kuma-website: app/_src/guides/consumer-producer-policies.md

---

> title: Namespace scoped policies

Reviewer: I don't have strong opinion here, but should the title mention "producer" and "consumer"? like "Producer and Consumer Policies"?

Author: changed

> ---

Author: added pitch

## Prerequisites

> - Completed [quickstart](/docs/{{ page.version }}/quickstart/kubernetes-demo/) to set up a zone control plane with demo application

Author: I am not sure if this should be right after quick start, I think we are missing basic policy guide right after quick start, and then we can move to this guide

```shell
echo "apiVersion: kuma.io/v1alpha1
kind: Mesh
metadata:
  name: default
spec:
  meshServices:
    mode: Exclusive
  mtls:
    enabledBackend: ca-1
    backends:
    - name: ca-1
      type: builtin" | kubectl apply -f -
```

To make sure that traffic works in our examples let's configure MeshTrafficPermission to allow all traffic:

```shell
echo "apiVersion: kuma.io/v1alpha1
kind: MeshTrafficPermission
metadata:
  namespace: kuma-system
  name: mtp
spec:
  targetRef:
    kind: Mesh
  from:
    - targetRef:
        kind: Mesh
      default:
        action: Allow" | kubectl apply -f -
```

To finish the setup we need to create two additional namespaces with sidecar injection for clients we will be using to communicate 
with our demo-app:

```shell
echo "apiVersion: v1
kind: Namespace
metadata:
  name: first-consumer
  labels:
    kuma.io/sidecar-injection: enabled
---
apiVersion: v1
kind: Namespace
metadata:
  name: second-consumer
  labels:
    kuma.io/sidecar-injection: enabled" | kubectl apply -f -
```

Now we can create deployment we will be using to communicate with our demo-app from a first-consumer namespace:

```shell

> kubectl run consumer --image nicolaka/netshoot -n first-consumer --command -- /bin/bash -c "while true; do ping localhost; sleep 60;done"

Author: makes sense, I've just copied the example from internet. Changed as suggested

```

## Namespace scoped policies

Now that we have our setup we can start playing with policies. Let's create simple MeshTimeout policy in `kuma-demo` namespace:

```shell
echo "apiVersion: kuma.io/v1alpha1
kind: MeshTimeout
metadata:
  name: producer-timeout
  namespace: kuma-demo
  labels:
    kuma.io/mesh: default
    kuma.io/origin: zone
spec:
  to:
    - targetRef:
        kind: MeshService
        name: demo-app
      default:
        http:
          requestTimeout: 1s" | kubectl apply -f -
```

We can now inspect applied policy labels:

```shell
kubectl get meshtimeout -n kuma-demo producer-timeout -o jsonpath='{.metadata.labels}'
```

We should see output like:

```json
{
  "k8s.kuma.io/namespace": "kuma-demo",
  "kuma.io/env": "kubernetes",
  "kuma.io/mesh": "default",
  "kuma.io/origin": "zone",
  "kuma.io/policy-role": "producer",
  "kuma.io/zone": "default"
}
```

{{site.mesh_product_name}} is adding custom labels to our policy. The one that interests us most at the moment is `kuma.io/policy-role`.
This is new label that indicates policy role. Possible values of this label are:

- `system` - policies applied in system namespace

> - `workload-owner` - policies that specify `spec.from` section

Author: changed

- `consumer` - policies targeting MeshServices from different namespace or targeting MeshService by labels

> - `producer` - policies targeting MeshServices from the same namespace

Reviewer: it's missing `to[]`, just "targeting MeshService" is not really clear even though you can't target it from anywhere except `to[]`

Author: changed

---
title: Producer and Consumer policies
---

> With namespace scoped policies in {{site.mesh_product_name}} you can have fine-grained control over policies and how they apply to 

Reviewer 2: We should mention empowering app owners by taking advantage of Kubernetes RBAC

Author: updated

```

To finish the setup we need to create two additional namespaces with sidecar injection for clients we will be using to communicate 
with our demo-app:

```shell
echo "apiVersion: v1
kind: Namespace
metadata:
  name: first-consumer
  labels:
    kuma.io/sidecar-injection: enabled
---
apiVersion: v1
kind: Namespace
metadata:
  name: second-consumer
  labels:
    kuma.io/sidecar-injection: enabled" | kubectl apply -f -
```

Now we can create deployment we will be using to communicate with our demo-app from a first-consumer namespace:

```shell
kubectl run consumer --image nicolaka/netshoot -n first-consumer --command -- /bin/bash -c "ping -i 60 localhost"
```

and from the second-consumer namespace:

```shell
kubectl run consumer --image nicolaka/netshoot -n second-consumer --command -- /bin/bash -c "ping -i 60 localhost"
```

You can make now make a couple of requests to our demo app to check if everything is working: 

```shell
kubectl exec -n first-consumer consumer -- curl -s -XPOST demo-app.kuma-demo:5000/increment
```

You should see something similar to:

```json
{"counter":"1","zone":"local","err":null}
```

## Namespace scoped policies

> Now that we have our setup we can start playing with policies. Let's create simple MeshTimeout policy in `kuma-demo` namespace:

Reviewer 3: Nit:

```suggestion
Now that we have our setup we can start playing with policies. Let's create a simple MeshTimeout policy in `kuma-demo` namespace:
```

> Now that we have our setup we can start playing with policies. Let's create a simple MeshTimeout policy in `kuma-demo` namespace:

Author: I think we should link to reference, added link

```shell
echo "apiVersion: kuma.io/v1alpha1
kind: MeshTimeout
metadata:
  name: producer-timeout
  namespace: kuma-demo
  labels:
    kuma.io/mesh: default
    kuma.io/origin: zone
spec:
  to:
    - targetRef:
        kind: MeshService

>         name: demo-app

Author: this is a producer policy so I will leave it as is, and update the consumer policy example

      default:
        http:
          requestTimeout: 1s" | kubectl apply -f -
```

We can now inspect applied policy labels:

```shell
kubectl get meshtimeout -n kuma-demo producer-timeout -o jsonpath='{.metadata.labels}'
```

We should see output like:

```json
{
  "k8s.kuma.io/namespace": "kuma-demo",
  "kuma.io/env": "kubernetes",
  "kuma.io/mesh": "default",
  "kuma.io/origin": "zone",
  "kuma.io/policy-role": "producer",
  "kuma.io/zone": "default"
}
```

{{site.mesh_product_name}} is adding custom labels to our policy. The one that interests us most at the moment is `kuma.io/policy-role`.

> This is new label that indicates policy role. Possible values of this label are:

Reviewer 4: ```suggestion
This label indicates the policy role. Possible values of this label are:
```

Anything `new` is very quickly old :)

- `system` - policies applied in system namespace
- `workload-owner` - policies that don't specify `spec.from` and `spec.to` sections or specify only `spec.from` section

> - `consumer` - policies targeting MeshServices from different namespace or targeting MeshService by labels

Reviewer 4: I think it's only policies with `spec.to`

Author: changed

> - `producer` - policies targeting MeshServices from the same namespace in `spec.to[]` section 

Reviewer 4: I think this needs a bit of rewording.

```suggestion
- `producer` - policies defined in the same namespace as the MeshService they are targeting in their `spec.to[]`. 
```

### Producer consumer model

> With namespace scoped policies we've introduced producer/consumer model for policies. 

Reviewer 2: ```suggestion
With namespace scoped policies we've introduced a producer/consumer model for policies. 
```

> **Producer** is someone who authors and operates service. Producer can create default policies that will be used when communicating with service. 

Reviewer 4: ```suggestion
**Producer** is someone who authors and operates a service. Producer can create default policies that will be used when communicating with service. 
```

> **Producer** is someone who authors and operates service. Producer can create default policies that will be used when communicating with service. 

Reviewer 2: ```suggestion
A **producer** is someone who authors and operates a service. A producer can create policies that will be applied by default to any communication with their services.
```

> Producer policies will be created in the same zone as MeshService they target. Producer policies will be synced to other zones. 

Reviewer 4: ```suggestion
Producer policies will be created in the same zone as `MeshService` they target. Producer policies will be synced to other zones. 
```

> **Consumer** is someone that utilizes a service by communicating with it. Consumer policies will be applied in consumer 

Reviewer 4: ```suggestion
**Consumer** is the client of a service. Consumer policies will be applied in consumer 
```

Reviewer 2: ```suggestion
A **consumer** is the client of a service. Consumer policies will be applied in the consumer
```

> namespace and will target MeshService from different namespace. Consumer policy will take effect only in consumer namespace.

Reviewer 4: ```suggestion
namespace and may target MeshService from different namespace. Consumer policy will take effect only in consumer namespace.
```

It can be in the same namespace (using labels).

> namespace and will target MeshService from different namespace. Consumer policy will take effect only in consumer namespace.

Reviewer 2: ```suggestion
namespace and may target MeshService from different namespaces. Consumer policies take effect only in the consumer namespace.
```

Consumer policies will override producer policies.

## Testing producer policy

To test MeshTimeout that we've applied in previous steps we need to simulate delays on our requests. To do this we need

> add header `x-set-response-delay-ms` to our requests

Reviewer 4: I'm curious if maybe it's better to do the things and then explain what they do rather than the other way round.

Author: I think it depends on a situation. In this case, it just felt natural to just explain it and then test it. But overall, we could try starting with example and then explanation for it

We can now make few requests to our demo-app, and we should see timeouts:

```shell
kubectl exec -n first-consumer consumer -- curl -s -XPOST demo-app.kuma-demo:5000/increment -H "x-set-response-delay-ms: 2000"
```

Example output:

```
{"counter":"3","zone":"local","err":null}
```

We should see the same results when making requests from second-consumer namespace:

```shell
kubectl exec -n second-consumer consumer -- curl -s -XPOST demo-app.kuma-demo:5000/increment -H "x-set-response-delay-ms: 2000"
```

## Utilizing consumer policy

> To this moment we were relying on producer policy created previously. Let's assume that we don't mind waiting a little longer

Reviewer 4: ```suggestion
Until now, we were relying on producer policy created previously. Let's assume that we don't mind waiting a little longer
```

for the response from demo-app, and we would like to override the timeout just for our namespace. To do so we need to create new policy:

```shell
echo "apiVersion: kuma.io/v1alpha1
kind: MeshTimeout
metadata:
  name: consumer-timeout
  namespace: first-consumer
  labels:
    kuma.io/mesh: default
    kuma.io/origin: zone
spec:
  to:
    - targetRef:
        kind: MeshService
        name: demo-app

>         namespace: kuma-demo

Reviewer 4: Same re `labels` it makes things a bit more obvious IMO

Author: changed

      default:
        http:
          requestTimeout: 3s" | kubectl apply -f -
```

When we now make requests from the first-consumer namespace all the requests should succeed, but they will take longer:

```shell
kubectl exec -n first-consumer consumer -- curl -s -XPOST demo-app.kuma-demo:5000/increment -H "x-set-response-delay-ms: 2000"
```

> Policy that we have just applied is consumer policy. This means that traffic from other namespaces to demo-app should not be affected by it.

Reviewer 4: ```suggestion
We have just applied a consumer policy. The timeout will only be applied in the `first-consumer` namespace.
```

We can test this by making requests from our second-consumer namespace:

```shell
kubectl exec -n second-consumer consumer -- curl -s -XPOST demo-app.kuma-demo:5000/increment -H "x-set-response-delay-ms: 2000"
```

We should still see timeouts:

```
upstream request timeout
```

## What we've learnt?

- How to apply policy on a namespace
- What are policy roles and what types can you encounter 
- What are producer and consumer policies
- How producer and consumer policies interact with each other

## Next steps

> - Read more about [producer/consumer policies](TODO)

Reviewer 4: It will be in:

```suggestion
- Read more about [producer/consumer policies](/docs/{{ page.version }}/policies/introduction)
```

Reviewer 4: Maybe it's worth cross linking to another guide like: `federate` no?

Author: linked federation guide

---
title: Namespace scoped policies
---

## Prerequisites

- Completed [quickstart](/docs/{{ page.version }}/quickstart/kubernetes-demo/) to set up a zone control plane with demo application

## Basic setup

> In order to be able to fully utilize namespace scoped policies you need to use real [MeshService](/docs/{{ page.version }}/networking/meshservice). 

Reviewer 4: ```suggestion
In order to be able to fully utilize namespace scoped policies you need to use [MeshService](/docs/{{ page.version }}/networking/meshservice). 
```

Author: changed

To enable MeshService resource generation on Mesh level we need to run:

```shell
echo "apiVersion: kuma.io/v1alpha1
kind: Mesh
metadata:
  name: default
spec:
  meshServices:

>     enabled: Everywhere

Author: you are right probably Exclusive is better option

  mtls:
    enabledBackend: ca-1
    backends:
    - name: ca-1
      type: builtin" | kubectl apply -f -
```

We also need HostnameGenerator for easy access to our services:

```shell
echo 'apiVersion: kuma.io/v1alpha1
kind: HostnameGenerator
metadata:
  name: all
  namespace: {{site.mesh_namespace}}
  labels:
    kuma.io/mesh: default
spec:
  selector:
    meshService:
      matchLabels:
        k8s.kuma.io/namespace: kuma-demo
  template: "{{ .DisplayName }}.{{ .Namespace }}.mesh"' | kubectl apply -f -

> ```

Reviewer 4: Isn't there already a hostnameGenerator now?

Author: we only have default hostNameGenerator for MES

metadata:
  namespace: {{site.mesh_namespace}}
  name: mtp
spec:
  targetRef:
    kind: Mesh
  from:
    - targetRef:
        kind: Mesh
      default:
        action: Allow" | kubectl apply -f -
```

To finish the setup we need to create two additional namespaces with sidecar injection for clients we will be using to communicate 
with our demo-app:

```shell
echo "apiVersion: v1
kind: Namespace
metadata:
  name: first-consumer
  labels:
    kuma.io/sidecar-injection: enabled
---
apiVersion: v1
kind: Namespace
metadata:
  name: second-consumer
  labels:
    kuma.io/sidecar-injection: enabled" | kubectl apply -f -
```

Now we can create deployment we will be using to communicate with our demo-app from a first-consumer namespace:

```shell
kubectl run consumer --image nicolaka/netshoot -n first-consumer --command -- /bin/bash -c "while true; do ping localhost; sleep 60;done"
```

and from the second-consumer namespace:

```shell
kubectl run consumer --image nicolaka/netshoot -n second-consumer --command -- /bin/bash -c "while true; do ping localhost; sleep 60;done"
```

You can make now make a couple of requests to our demo app to check if everything is working: 

```shell

> kubectl exec -n first-consumer consumer -- curl --no-progress-meter demo-app.kuma-demo.mesh:5000/counter

Reviewer 4: I'd use `kubectl exec -n first-consumer consumer -- curl -s -XPOST demo-app.kuma-demo:5000/increment` 

A few things different here:
- use the k8s hostname because that enables the guide to introduce less information
- use `-s` instead of `--no-progress-meter` it's quieter
- use `-X POST /increment` I find it more straightforward when calls do something.

Author: makes sense, changed as requested

Now that we have our setup we can start playing with policies. Let's create simple MeshTimeout policy in `kuma-demo` namespace:

```shell
echo "apiVersion: kuma.io/v1alpha1
kind: MeshTimeout
metadata:
  name: producer-timeout
  namespace: kuma-demo
  labels:
    kuma.io/mesh: default
    kuma.io/origin: zone
spec:
  to:
    - targetRef:
        kind: MeshService
        name: demo-app
      default:
        http:
          requestTimeout: 1s" | kubectl apply -f -
```

We can now inspect applied policy labels:

```shell
kubectl get meshtimeout -n kuma-demo producer-timeout -o jsonpath='{.metadata.labels}'
```

We should see output like:

```json
{
  "k8s.kuma.io/namespace": "kuma-demo",
  "kuma.io/env": "kubernetes",
  "kuma.io/mesh": "default",
  "kuma.io/origin": "zone",
  "kuma.io/policy-role": "producer",
  "kuma.io/zone": "default"
}
```

{{site.mesh_product_name}} is adding custom labels to our policy. The one that interests us most at the moment is `kuma.io/policy-role`.
This is new label that indicates policy role. Possible values of this label are:

- `system` - policies applied in system namespace
- `workload-owner` - policies that specify `spec.from` section

> - `consumer` - policies targeting MeshServices from different namespace

Reviewer 4: Actually that's not exactly true and it was my confusing in the issue: https://github.com/kumahq/kuma/issues/11570

I think to make it clear to user we need to rework this definition

Author: right, since we are not resolving labels to actual MeshServices when setting policy-role label this ends up being a consumer policy. Shouldn't `policy-role` be set to `producer` only if targeted MeshService is in the same namespace as policy and topLevel targetRef is `Mesh`?

- `producer` - policies targeting MeshServices from the same namespace

### Producer consumer model

With namespace scoped policies we've introduced producer/consumer model for policies. 

**Producer** is someone who authors and operates service. Producer can create default policies that will be used when communicating with service. 
Producer policies will be created in the same zone as MeshService they target. Producer policies will be synced to other zones. 

**Consumer** is someone that utilizes a service by communicating with it. Consumer policies will be applied in consumer 
namespace and will target MeshService from different namespace. Consumer policy will take effect only in consumer namespace.

> Consumer policies will override producer policies.

Reviewer 4: Does it make sense to explain with a mermaid diagram which sidecars is getting updated in either case?

Author: I don't this there is good mermaid diagram type to show this. But I was playing with making a diagram that would make sense and I came up with this: 
![Screenshot 2024-10-01 at 11 55 42](https://github.com/user-attachments/assets/b16f97cc-ce5b-4532-8977-363e1d5cee3a)
Let me know what you think about it

Reviewer: Somehow I feel uncomfortable seeing server on the left and client on the right 😄 IDK is it just me or `Client --> Server` has more familiar look to it?

Reviewer 4: You can use mermaid subgraph: https://mermaid.js.org/syntax/flowchart.html#subgraphs

Reviewer 4: Here's an example of a draft I was working on for something else:

```mermaid
flowchart LR

clients@{ shape: processes, label: "clients" }
servers@{ shape: processes, label: "servers"}

subgraph Data-plane
   app
   subgraph data-plane proxy/Envoy
   inbounds
   outbounds
   end
end

inbounds -.local traffic.-> app
app -.local traffic.-> outbounds

clients --> inbounds
outbounds --> servers
```

I think you could show your setup with your app and the 2 different consumers and show how producer/consumer policies affect different sutff (the same base graph would evolve progressively).

Author: added diagrams

## Testing producer policy

To test MeshTimeout that we've applied in previous steps we need to simulate delays on our requests. To do this we need
MeshFaultInjection policy

```shell
echo "apiVersion: kuma.io/v1alpha1
kind: MeshFaultInjection
metadata:
  name: default
  namespace: kuma-demo
  labels:
    kuma.io/mesh: default
spec:
  targetRef:
    kind: Mesh
    proxyTypes: [Sidecar]
  from:
    - targetRef:
        kind: Mesh
      default:
        http:
          - delay:
              percentage: '50'
              value: 2s" | kubectl apply -f -

> ```

Reviewer 4: I'm on the fence with using the Mesh to test the Mesh that's also something @johnharris85 found it confused users.

Maybe we should just add an option to delay the response in the demo app?

Author: Yeah, I not a big fan of testing this with fault injection but this is what we did previously. I've added possibility to specify delay and specific status code in kuma-counter-demo: https://github.com/kumahq/kuma-counter-demo/pull/53

for the response from demo-app, and we would like to override the timeout just for our namespace. To do so we need to create new policy:

```shell
echo "apiVersion: kuma.io/v1alpha1
kind: MeshTimeout
metadata:
  name: consumer-timeout
  namespace: first-consumer
  labels:
    kuma.io/mesh: default
    kuma.io/origin: zone
spec:
  to:
    - targetRef:
        kind: MeshService
        name: demo-app
        namespace: kuma-demo
      default:
        http:
          requestTimeout: 3s" | kubectl apply -f -
```

When we now make requests from the first-consumer namespace all the requests should succeed, but they will take longer:

```shell
kubectl exec -n first-consumer consumer -- curl --no-progress-meter demo-app.kuma-demo.mesh:5000/counter
```

Policy that we have just applied is consumer policy. This means that traffic from other namespaces to demo-app should not be affected by it.
We can test this by making requests from our second-consumer namespace:

```shell
kubectl exec -n second-consumer consumer -- curl --no-progress-meter demo-app.kuma-demo.mesh:5000/counter
```

We should still see timeouts:

```
upstream request timeout

> ```

Reviewer 4: @slonka in his latest guide has added a `Things to remember section` maybe a general `What we've learnt?` would be useful here?

Author: added `What we've learnt?` section
